Account for rounding errors when deciding the diagnostic boundaries

estebank opened the pull request:
Fix Miri by fixing the bug raised in https://github.com/rust-lang/rust/pull/63402#discussion_r319485328.

Fixes https://github.com/rust-lang/rust/issues/64020

estebank committed:
Account for rounding errors when deciding the diagnostic boundaries

rust-highfive commented:
r? @zackmdavis

(rust_highfive has picked a reviewer for you, use r? to override)

estebank commented:
<img width="837" alt="Screen Shot 2019-08-30 at 5 47 08 PM" src="https://user-images.githubusercontent.com/1606434/64057142-3ba3f800-cb4e-11e9-9628-2f8346c6ab21.png">
<img width="556" alt="Screen Shot 2019-08-30 at 5 46 53 PM" src="https://user-images.githubusercontent.com/1606434/64057145-4068ac00-cb4e-11e9-9e17-303f1b1fe4b5.png">

CC @RalfJung

estebank committed:
Tweak terminal width trimming

Properly account for left margin when setting terminal width through
CLI flag and don't trim code by default if we can't get the terminal's
dimensions.

RalfJung commented on src/librustc_errors/emitter.rs:
```diff
@@ -146,12 +146,12 @@ impl Margin {
                 // Attempt to fit the code window considering only the spans and labels.
                 let padding_left = (self.column_width - (self.label_right - self.span_left)) / 2;
-                self.computed_left = self.span_left - padding_left;
+                self.computed_left = max(self.span_left, padding_left) - padding_left;
```
These could also all use `saturating_sub` instead, right? What do you think would be more readable?

RalfJung commented:
Also I wonder, is there any way you could add a test for this? (But IMO that's not a blocker.)

RalfJung commented:
I can confirm that this fixes Miri. r=me with the comment about `saturating_sub` resolved (one way or another).

estebank committed:
Use saturating_sub

estebank commented:
> Also I wonder, is there any way you could add a test for this?

I haven't found why the `compile-flags` doesn't propagate `-Zterminal-width` properly to tests. I'll get back to it to increase test coverage, but for now I'm letting it slide with some manual testing 😬

estebank commented:
r? @RalfJung @bors r=RalfJung rollup

bors commented:
:pushpin: Commit 84567190e0d36f9a61f9bc833bd9fa559aeb0089 has been approved by `RalfJung`

<!-- @bors r=RalfJung 84567190e0d36f9a61f9bc833bd9fa559aeb0089 -->
<!-- homu: {"type":"Approved","sha":"84567190e0d36f9a61f9bc833bd9fa559aeb0089","approver":"RalfJung"} -->

bjorn3 commented on src/librustc_errors/emitter.rs:
```diff
@@ -1304,11 +1304,13 @@ impl EmitterWriter {
-                    term_size::dimensions().map(|(w, _)| w - code_offset).unwrap_or(140)
+                    term_size::dimensions()
+                        .map(|(w, _)| w.saturating_sub(code_offset))
+                        .unwrap_or(std::usize::MAX)
```
Why was this changed to usize::MAX?

estebank replied:
@bjorn3 If the dimensions aren't supplied and can't be queried, I decided that keeping the _current_ uncropped behavior makes more sense. By setting it to `MAX` it'll never crop.
